stdlib/public/Concurrency/Actor.cpp

```diff
@@ -935,356 +1034,370 @@ static void traceJobQueue(DefaultActorImpl *actor, Job *first) {
-                         oldState.Flags.getOpaqueValue(),                      \
-                         newState.Flags.getOpaqueValue(), __FUNCTION__);       \
-    concurrency::trace::actor_state_changed(                                   \
-        this, newState.FirstJob.getRawJob(),                                   \
-        newState.FirstJob.needsPreprocessing(),                                \
-        newState.Flags.getOpaqueValue());                                      \
-  } while (0)
-    ACTOR_STATE_TRANSITION;
-
-    // The priority of the remaining work.
-    auto newPriority = newState.Flags.getMaxPriority();
-
-    // This is the actor's owning job; per the ownership rules (see
-    // the comment on DefaultActorImpl), if there are remaining
-    // jobs, we need to balance out our ownership one way or another.
-    // We also, of course, need to ensure that there's a thread that's
-    // actually going to process the actor.
-    if (hasMoreJobs) {
-      // Enqueue a job that will try to take over running with the
-      // new priority.  This also ensures that there's a job at that
-      // priority which will actually take over the actor.
-      scheduleNonOverrideProcessJob(newPriority, hasActiveInlineJob);
+bool DefaultActorImpl::tryLock(bool asDrainer) {
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+  SWIFT_TASK_DEBUG_LOG("Thread %#x attempting to jump onto %p, as drainer = %d", dispatch_lock_value_for_self(), this, asDrainer);
+  dispatch_thread_override_info_s threadOverrideInfo;
+  threadOverrideInfo = swift_dispatch_thread_get_current_override_qos_floor();
+  qos_class_t overrideFloor = threadOverrideInfo.override_qos_floor;
+
+retry:;
+#else
+  SWIFT_TASK_DEBUG_LOG("Thread attempting to jump onto %p, as drainer = %d", this, asDrainer);
+#endif
+
+  auto oldState = CurrentState.load(std::memory_order_relaxed);
+  while (true) {
+
+    if (asDrainer) {
+      // TODO (rokhinip): Once we have OOL process job support, this assert can
+      // potentially fail due to a race with an actor stealer that might have
+      // won and started running the task
+      assert(oldState.isScheduled());
+
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+      // We only want to self override a thread if we are taking the actor lock
+      // as a drainer because there might have been higher priority work
+      // enqueued that might have escalated the max priority of the actor to be
+      // higher than the original thread request.
+      qos_class_t maxActorPriority = (qos_class_t) oldState.getMaxPriority();
+
+      if (threadOverrideInfo.can_override && (maxActorPriority > overrideFloor)) {
+        SWIFT_TASK_DEBUG_LOG("[Override] Self-override thread with oq_floor %#x to match max actor %p's priority %#x", overrideFloor, this, maxActorPriority);
+
+        (void) swift_dispatch_thread_override_self(maxActorPriority);
+        overrideFloor = maxActorPriority;
+        goto retry;
```

Comment:
Why does this `goto retry` instead of `continue`?

Reply:
Seems like that would be rare, so it would be better to go through and fail in the CAS. Easier to understand as well. It's OK with me if you want to keep it the way it is, though, with a comment capturing what you just said here.

Reply:
If we have an actor with heavy contention, it's quite likely we'd end up in this case. I'll add comments about this to make it clear.

I kept the logic similar to what dispatch does today: https://github.com/apple/swift-corelibs-libdispatch/blob/main/src/inline_internal.h#L1235

Reply:
Sounds good. I trust your judgment as to what's rare/common here much more than my own.